[강병훈] sprint4

kbh-n opened the pull request:
## 요구사항

### 기본
- [x] 기본 항목 1
- [x] 기본 항목 2

### 심화
- [x] 심화 항목 1
- [x] 심화 항목 2

## 주요 변경사항
- 
- 



## 스크린샷
<img width="1235" alt="스크린샷 2025-02-18 오후 1 29 54" src="https://github.com/user-attachments/assets/7ce210f9-8825-4472-b6ba-18cc8958e47f" />




## Postman API 테스트 결과
[스프린트4.fianl.json](https://github.com/user-attachments/files/18896763/4.fianl.json)



## 멘토에게
- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.
-

kbh-n committed:
update

kbh-n committed:
build.gradle 업데이트

kbh-n committed:
File*Repository 리펙터링, Spring 오류 수정

kbh-n committed:
UserStatus 리팩토링

kbh-n committed:
entity 리팩토링, login & userStatusUpdate 기능 구현

kbh-n committed:
채널관리 구현

kbh-n committed:
메시지 관리 구현

kbh-n committed:
메시지  수신 정보 관리 구현

kbh-n committed:
바이너리 파일 다운로드 구현

kbh-n committed:
심화 요구 사항 구현 및 리펙토링

kbh-n committed:
refector: build.gradle

kbh-n committed:
messageRepo 리펙토링

kbh-n committed:
RestFul refact

kbh-n committed:
This reverts commit d48100d38c1ee7ebfc801c012304ef60b2521f60.

kbh-n committed:
conflict test: 1

itstimi-XD committed:
Fix: 모든 Merge Conflict ours로 해결

itstimi-XD commented:
### 🔹 Postman API 테스트 응답 결과 확인

~~⚠️ Postman 요청이 잘 정의되어 있지만, 응답 결과(JSON)가 올바르게 출력되지 않는 것으로 보입니다.~~
- 응답이 비어 있는([])경우 API가 정상적으로 동작하지 않을 가능성이 있습니다. 
- API가 정상 동작하는지, 예외 발생 시 적절한 응답이 오는지 확인이 필요합니다.

📌 개선 제안:
~~ - Postman에서 요청을 다시 실행한 뒤 응답데이터가 나온 후에 export 시도해봐주세요.~~

❗ API 테스트는 요청만 있는 것이 아니라, 실제 응답이 올바르게 오는지까지 검증하는 과정이 중요합니다.
~~이 부분을 보완해서 다시 확인 부탁드립니다! 🚀~~


저도 해보니 응답은 포함이 안되네요🤔

itstimi-XD committed:
⚙️ .gitignore 수정: 불필요한 파일 제외 오류 수정

이전에 제가 코드 리뷰를 잘못해서 .gitignore에서 빌드 관련 파일이 잘못 제외되지 않도록 설정해주셨었네요.
그로 인해 불필요한 파일이 Git에 포함되었기에, 이를 수정하여 올바르게 제외 처리했습니다.

- gradle-wrapper.jar가 제외되지 않도록 수정
- build/, out/ 등의 불필요한 파일 올바르게 제외
- Mac OS 관련 불필요한 파일 추가 제외

itstimi-XD commented:
### 💡 .gitignore 수정 사항 안내

제 기억으로는 제가 제외시키지 않아도 되는 파일을 제외시켜달라고 말씀드렸던 분들이 계셨던 것 같아..병훈님 코드 Conflict 해결 겸 다운로드 받은 김에 일부 수정해서 올려드려요!

✅ 주요 수정 내용:
- gradle-wrapper.jar가 제외되지 않도록 설정 -> 이 부분이 제가 잘못 말씀드렸..거나? 병훈님이 실수하셨거나 했던 것 같아요~!
- build/, out/ 등 불필요한 파일이 올바르게 .gitignore에서 제외됨
- Mac OS에서 자동 생성되는 불필요한 파일 (.DS_Store, ._*) 추가 제외

⚠️ Pull 받고 진행해주세요!
이 변경 사항을 적용하려면 PR 머지 후 pull 한 뒤 진행해 주세요.
기존에 Git에 올라간 불필요한 파일들이 있다면, 아래 명령어를 실행한 후 다시 커밋 & 푸시하면 됩니다.
```
git rm -r --cached .
git add .
git commit -m "불필요한 파일 제거 및 .gitignore 반영"
git push origin <your-branch-name>
```

itstimi-XD commented on 1-sprint-mission/src/main/java/com/sprint/mission/discodeit/controller/channel/ChannelController.java:
```diff
@@ -0,0 +1,54 @@
+    this.channelService = channelService;
+  }
+
+  //공개 채널 생성
+  @PostMapping("/public")
+  public ChannelResponseDTO createChannel(@RequestBody PublicChannelCreateDTO createDTO) {
+    return channelService.createPublicChannel(createDTO);
+  }
+
+  //비공개 채널 생성
+  @PostMapping("/private")
+  public ChannelResponseDTO createPrivateChannel(@RequestBody PrivateChannelCreateDTO createDTO) {
+    return channelService.createPrivateChannel(createDTO);
+  }
+
+  //공개 채널의 정보를 수정
+  @PatchMapping("/public")
+  public ChannelResponseDTO updatePublicChannel(@RequestBody ChannelUpdateDTO updateDTO) {
+    return channelService.update(updateDTO);
+  }
+
```
이렇게 공개, 비공개 채널을 나눠서 요청받을 수도 있지만 하나의 생성, 수정 endpoint로 할 수도 있어요.
요청DTO에서 public/private 여부를 받아서 처리하는거죠!

itstimi-XD commented on 1-sprint-mission/src/main/java/com/sprint/mission/discodeit/service/BinaryContentService.java:
현재 해당 코드에서 예외 발생 시에 IllegalArgumentException를 던지는데요, 제 생각에는 파일과 관련된 예외 클래스 정도는 따로 만드는게 어떨까 싶어요. 
BinaryContentException이라던지요,,!

itstimi-XD replied:
이전 변경사항이긴 하지만, 해당 코드의 
```java
System.out.println("BinaryContent deleted: " + binaryContentId);
log.error("BinaryContent deleted: {}", binaryContentId);
```

이 부분은 

```java
log.info("BinaryContent deleted: {}", binaryContentId);
```
로 수정하시는 것이 좀 더 좋다고 생각해요. 

- log.error()는 오류 발생 시 쓰는 게 좋으므로, 정상적인 삭제는 log.info()가 적절해요.
- log.warn()은 위험 가능성이 있는 경우에 사용합니다.

itstimi-XD replied:
또한 findAllByIdIn(List<UUID> binaryContentIds) 메서드에서도 

❌ 문제점
- 재 findAllByIdIn()에서는 List<UUID>를 받아서 .stream()을 사용하여 변환하는데,
만약 binaryContentIds가 빈 리스트일 경우 불필요한 DB 호출이 발생할 가능성이 있어요.
✅ 해결 방법
- 빈 리스트라면 바로 Collections.emptyList() 반환하는 게 좋아요.
```java
public List<BinaryContentResponseDTO> findAllByIdIn(List<UUID> binaryContentIds){
    if (binaryContentIds == null || binaryContentIds.isEmpty()) {
        return Collections.emptyList();
    }
    return binaryContentRepository.findAllByIdIn(binaryContentIds).stream()
            .map(BinaryContentResponseDTO::fromEntity)
            .toList();
}
```

itstimi-XD commented on 1-sprint-mission/src/main/java/com/sprint/mission/discodeit/service/basic/BasicChannelService.java:
```diff
@@ -27,163 +27,167 @@
-
+    for (User member : members) {
+      ReadStatus readStatus = new ReadStatus(member, privateChannel, Instant.now());
+      readStatusRepository.save(readStatus); // 레포지토리 구현 해야함 확인!
```
인텔리제이를 쓰고 계신다면 
// TODO : 해야할 일 

이렇게 써보시면 IDE에서 편하게 확인이 가능해요.

itstimi-XD commented on 1-sprint-mission/src/main/java/com/sprint/mission/discodeit/service/basic/BasicChannelService.java:
❌ 문제점
- N+1 문제 발생 가능성:
현재 findAllByUserId()에서 유저가 속한 채널을 조회할 때 여러 번 findById()를 호출하는 방식으로 동작
- 중복된 채널 제거 로직이 비효율적:
HashSet<Channel>을 사용해서 중복 제거하고 있지만, 쿼리를 여러 번 호출하는 구조

✅ 개선 방법
- 한 번의 JOIN FETCH 쿼리로 유저가 속한 모든 채널을 가져오도록 개선할 수 있음.
- @Query를 활용해서 최적화하면 SQL 호출 횟수를 줄일 수 있음.
```java
@Query("SELECT c FROM Channel c LEFT JOIN FETCH c.readStatuses rs WHERE rs.user.id = :userId OR c.owner.id = :userId")
List<Channel> findAllByUserId(@Param("userId") UUID userId);
```
주석에 써주신 것처럼 성능이 고민된다면 이 방법을 사용해볼 수도 있어요!

itstimi-XD commented on 1-sprint-mission/src/main/java/com/sprint/mission/discodeit/service/basic/BasicMessageService.java:
```diff
@@ -30,7 +31,7 @@ public class BasicMessageService implements MessageService {
 
     @Override
-    public MessageResponseDTO create(MessageCreateDTO messageCreateDTO) {
+    public MessageResponseDTO create(MessageCreateDTO messageCreateDTO, BinaryContentCreateRequest binaryContentCreateRequest) {
```
해당 메서드에는 @Transactional 어노테이션 추가하는 것을 고민해보세요.
